### PR TITLE
test(semver): add test for parse_range

### PR DESCRIPTION
**Pull request by javihernant**

### CODE REVIEW

**Comment by javihernant** [comment]
This is just a draft. I am not sure if test should be implemented like this. Can anyone confirm?

File: semver/parse_range_test.ts [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+  }
+  for (let i = 0; i < cs0.length; i++) {
+    if (!equalComp(cs0[i], cs1[i])) {
+      return false;
+    }
+  }
+  return true;
+}
+
+function equalRanges(
```
Comment (kt3k) [inline semver/parse_range_test.ts:34]: I don't think we need this util. We can use `assertEquals` to compare nested objects.

**Comment by javihernant** [comment]
This new commit contains one way to test `parseRange()`. The only problem I see here is, if parseRange() implementation is slightly changed so that the object returned is a bit different from that in the expected array (in `parse_args_test.ts`), then the tests are going to fail. For example, when I originally made the `parse_args_test.ts` file, all tests worked fine. However, today I have rechecked and one of them didn't work because one of the objects returned had changed. It changed from having the property "operator" as an empty string to an undefined value. How do you think we can tackle this issue? What changes can I make to this PR?

**Comment by kt3k** [comment]
> The only problem I see here is, if parseRange() implementation is slightly changed so that the object returned is a bit different from that in the expected array (in parse_args_test.ts), then the tests are going to fail. For example, when I originally made the parse_args_test.ts file, all tests worked fine. However, today I have rechecked and one of them didn't work because one of the objects returned had changed. It changed from having the property "operator" as an empty string to an undefined value. How do you think we can tackle this issue?

I think that's just fine. Tests are for catching such changes. We'll update the test expectation in that cases if the change looks reasonable.

File: semver/parse_range_test.ts [hunk 2]
```diff
@@ -0,0 +1,869 @@
...
+    ]],
+  ];
+
+  for (const [r, expected] of ranges) {
+    const range = parseRange(r);
+    assertEquals(range, expected, 'call to parseRange("' + r + '") failed');
+  }
+});
+
+Deno.test("parseHyphenRanges", () => {
```
Comment (kt3k) [inline semver/parse_range_test.ts:569]: We have some rule for naming test cases: https://github.com/denoland/deno_std/blob/main/.github/CONTRIBUTING.md#tests

This case probably should be something like:
```suggestion
Deno.test("parseRange() parses ranges with hyphens", () => {
```

Can you also update the other test cases alike?

Comment (javihernant) [reply semver/parse_range_test.ts]: done :)

File: semver/parse_range_test.ts [hunk 3]
```diff
@@ -0,0 +1,869 @@
...
+        { operator: ">=", major: 1, minor: 2, patch: 3, prerelease: [] },
+        { operator: "<", major: 2, minor: 0, patch: 0 },
+      ],
+    ]],
+    ["^1.2.3+build", [
+      [
+        { operator: ">=", major: 1, minor: 2, patch: 3, prerelease: [] },
+        { operator: "<", major: 2, minor: 0, patch: 0 },
+      ],
+    ]],
```
Comment (kt3k) [inline semver/parse_range_test.ts:60]: This one seems the same as the above one?

File: semver/parse_range_test.ts [hunk 4]
```diff
@@ -0,0 +1,869 @@
...
+          operator: "<=",
+          major: 1,
+          minor: 2,
+          patch: 3,
+          prerelease: [],
+          build: [],
+        },
+      ],
+    ]],
+    ["^1.2.3", [
```
Comment (kt3k) [inline semver/parse_range_test.ts:106]: This case seems appearing twice

File: semver/parse_range_test.ts [hunk 5]
```diff
@@ -0,0 +1,869 @@
...
+          patch: 3,
+          prerelease: [],
+          build: [],
+        },
+      ],
+    ]],
+    ["=1.2.3", [
+      [
+        {
+          operator: undefined,
```
Comment (kt3k) [inline semver/parse_range_test.ts:127]: I wonder why this isn't `"="`? 🤔

File: semver/parse_range_test.ts [hunk 6]
```diff
@@ -0,0 +1,869 @@
...
+          operator: ">",
+          major: 1,
+          minor: 0,
+          patch: 0,
+          prerelease: [],
+          build: [],
+        },
+      ],
+    ]],
+    [">1.0.0", [
```
Comment (kt3k) [inline semver/parse_range_test.ts:203]: Looks a duplicate

**Comment by kt3k** [comment]
@javihernant Thanks for updating! Looks close to be merged, but I found several cases were repeated with completely the same input and expectations. Can you clean up them?

File: semver/parse_range_test.ts [hunk 7]
```diff
@@ -0,0 +1,869 @@
...
+          operator: "<=",
+          major: 2,
+          minor: 0,
+          patch: 0,
+          prerelease: [],
+          build: [],
+        },
+      ],
+    ]],
+    ["<=2.0.0", [
```
Comment (kt3k) [inline semver/parse_range_test.ts:227]: seems a duplicate

File: semver/parse_range_test.ts [hunk 8]
```diff
@@ -0,0 +1,869 @@
...
+          operator: "<",
+          major: 0,
+          minor: 0,
+          patch: 1,
+          prerelease: [],
+          build: [],
+        },
+      ],
+    ]],
+    [">=0.2.3 || <0.0.1", [
```
Comment (kt3k) [inline semver/parse_range_test.ts:331]: seems a duplicate

File: semver/parse_range_test.ts [hunk 9]
```diff
@@ -0,0 +1,869 @@
...
+      [
+        { operator: ">=", major: 1, minor: 2, patch: 0 },
+        { operator: "<", major: 1, minor: 3, patch: 0 },
+      ],
+      [
+        { operator: ">=", major: 2, minor: 0, patch: 0 },
+        { operator: "<", major: 3, minor: 0, patch: 0 },
+      ],
+    ]],
+    ["1.2.x || 2.x", [
```
Comment (kt3k) [inline semver/parse_range_test.ts:375]: seems a duplicate

File: semver/parse_range_test.ts [hunk 10]
```diff
@@ -0,0 +1,869 @@
...
+        { operator: "<", major: 3, minor: 0, patch: 0 },
+      ],
+    ]],
+    ["2.*.*", [
+      [
+        { operator: ">=", major: 2, minor: 0, patch: 0 },
+        { operator: "<", major: 3, minor: 0, patch: 0 },
+      ],
+    ]],
+    ["2.*.*", [
```
Comment (kt3k) [inline semver/parse_range_test.ts:391]: seems a duplicate

File: semver/parse_range_test.ts [hunk 11]
```diff
@@ -0,0 +1,869 @@
...
+      [
+        { operator: ">=", major: 1, minor: 2, patch: 0 },
+        { operator: "<", major: 1, minor: 3, patch: 0 },
+      ],
+      [
+        { operator: ">=", major: 2, minor: 0, patch: 0 },
+        { operator: "<", major: 3, minor: 0, patch: 0 },
+      ],
+    ]],
+    ["1.2.* || 2.*", [
```
Comment (kt3k) [inline semver/parse_range_test.ts:413]: seems a duplicate

File: semver/parse_range_test.ts [hunk 12]
```diff
@@ -0,0 +1,869 @@
...
+        { operator: "<", major: 0, minor: 1, patch: 0 },
+      ],
+    ]],
+    ["~2.4", [
+      [
+        { operator: ">=", major: 2, minor: 4, patch: 0 },
+        { operator: "<", major: 2, minor: 5, patch: 0 },
+      ],
+    ]],
+    ["~2.4", [
```
Comment (kt3k) [inline semver/parse_range_test.ts:453]: seems a duplicate

File: semver/parse_range_test.ts [hunk 13]
```diff
@@ -0,0 +1,869 @@
...
+        { operator: "<", major: 2, minor: 5, patch: 0 },
+      ],
+    ]],
+    ["~>3.2.1", [
+      [
+        { operator: ">=", major: 3, minor: 2, patch: 1, prerelease: [] },
+        { operator: "<", major: 3, minor: 3, patch: 0 },
+      ],
+    ]],
+    ["~>3.2.1", [
```
Comment (kt3k) [inline semver/parse_range_test.ts:465]: seems a duplicate

File: semver/parse_range_test.ts [hunk 14]
```diff
@@ -0,0 +1,869 @@
...
+        { operator: "<", major: 3, minor: 3, patch: 0 },
+      ],
+    ]],
+    ["~1", [
+      [
+        { operator: ">=", major: 1, minor: 0, patch: 0 },
+        { operator: "<", major: 2, minor: 0, patch: 0 },
+      ],
+    ]],
+    ["~>1", [
```
Comment (kt3k) [inline semver/parse_range_test.ts:477]: seems a duplicate

File: semver/parse_range_test.ts [hunk 15]
```diff
@@ -0,0 +1,869 @@
...
+    [">=0.7.x", [[{ operator: ">=", major: 0, minor: 7, patch: 0 }]]],
+    ["<0.7.x", [[{ operator: "<", major: 0, minor: 7, patch: 0 }]]],
+    [">1.2", [[{ operator: ">=", major: 1, minor: 3, patch: 0 }]]],
+    ["^0.0.1", [
+      [
+        { operator: ">=", major: 0, minor: 0, patch: 1, prerelease: [] },
+        { operator: "<", major: 0, minor: 0, patch: 2 },
+      ],
+    ]],
+    ["^0.0.1", [
```
Comment (kt3k) [inline semver/parse_range_test.ts:518]: seems a duplicate

File: semver/parse_range_test.ts [hunk 16]
```diff
@@ -0,0 +1,869 @@
...
+      [
+        { operator: ">=", major: 0, minor: 0, patch: 1, prerelease: [] },
+        { operator: "<", major: 0, minor: 0, patch: 2 },
+      ],
+    ]],
+    ["^1.2.3", [
```
Comment (kt3k) [inline semver/parse_range_test.ts:524]: seems a duplicate

File: semver/parse_range_test.ts [hunk 17]
```diff
@@ -0,0 +1,869 @@
...
+        { operator: "<", major: 2, minor: 0, patch: 0 },
+      ],
+    ]],
+    ["^1.2.3", [
+      [
+        { operator: ">=", major: 1, minor: 2, patch: 3, prerelease: [] },
+        { operator: "<", major: 2, minor: 0, patch: 0 },
+      ],
+    ]],
+    ["^1.2", [
```
Comment (kt3k) [inline semver/parse_range_test.ts:536]: seems a duplicate

File: semver/parse_range_test.ts [hunk 18]
```diff
@@ -0,0 +1,869 @@
...
+          major: 0,
+          minor: 0,
+          patch: 0,
+          prerelease: [],
+          build: [],
+          semver: { major: 0, minor: 0, patch: 0, prerelease: [], build: [] },
+        },
+      ],
+    ]],
+    ["^1.2.3", [
```
Comment (kt3k) [inline semver/parse_range_test.ts:555]: seems a duplicate

**Comment by javihernant** [comment]
Updated the tests to remove all duplicates. I'm sorry I left so many!!!

**Review by kt3k — APPROVED** [review]
LGTM. Thanks for working on this!

**Pull request merged**